
MVP 2 - João Pedro Cunha - Julho 2024 


Detalhamento e objetivos


Este projeto consiste em analisar dados da temporada regular e de playoffs da NBA 2021-2022. Oss datasets utilizados foram baixados diretamente do Kaggle conforme o link https://www.kaggle.com/datasets/vivovinco/nba-player-stats/data e não há licença de uso para os dados escolhidos. Os dados consistem em dois arquivos .csv com as estatísticas individuais dos jogadores da liga, sendo um para a fase regular e outro para a fase de playoffs.

Este trabalho se propõe a responder as perguntas abaixo e analisar como a resposta muda em relação as diferentes fases (regular versus playoffs).

1) Como as estatísticas ofensivas (pontos, assistências, número de arremessos) se relacionam com as estatísticas defensivas (roubos de bola, bloqueios) dos jogadores ? Esta informação é extremamente útil para a comissão técnica dos times envolvidos na liga.

2) Quais os jogadores que mais se destacam tanto ofensivamente quanto defensivamente ? Por sua vez, esta informação é extremamente útil para os managers das equipes da liga já visando a janela de transferência de jogadores para a temporada 2022-2023.

3) O desempenho dos jogadores muda com a idade ? Esta informação é extremamente pertinente para pautar uma discussão entre jornalistas durante a transmissão de um programa de televisão esportivo que aborde a NBA durante sua grade.

4) Considerando a temporada completa, quais foram os três melhores jogadores da liga ?


Coleta e armazenamento de dados na nuvem no modelo flat

In [0]:
# Importando dados da temporada regular

# Localização e tipo do arquivo
file_location = "/FileStore/tables/2021_2022_NBA_Player_Stats___Regular.csv"
file_type = "csv"

# Parâmetros do arquivo .csv
infer_schema = "false"
first_row_is_header = "true"
delimiter = ","

# Criando dataframe da temporada regular
df = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", ";") \
  .load(file_location)

display(df.head(10))

Rk Player Pos Age Tm G GS MP FG FGA FG% 3P 3PA 3P% 2P 2PA 2P% eFG% FT FTA FT% ORB DRB TRB AST STL BLK TOV PF PTS 1 Precious Achiuwa C 22 TOR 73 28 23.6 3.6 8.3 0.439 0.8 2.1 0.359 2.9 6.1 0.468 0.486 1.1 1.8 0.595 2 4.5 6.5 1.1 0.5 0.6 1.2 2.1 9.1 2 Steven Adams C 28 MEM 76 75 26.3 2.8 5.1 0.547 0 0 0 2.8 5 0.548 0.547 1.4 2.6 0.543 4.6 5.4 10 3.4 0.9 0.8 1.5 2 6.9 3 Bam Adebayo C 24 MIA 56 56 32.6 7.3 13 0.557 0 0.1 0 7.3 12.9 0.562 0.557 4.6 6.1 0.753 2.4 7.6 10.1 3.4 1.4 0.8 2.6 3.1 19.1 4 Santi Aldama PF 21 MEM 32 0 11.3 1.7 4.1 0.402 0.2 1.5 0.125 1.5 2.6 0.56 0.424 0.6 1 0.625 1 1.7 2.7 0.7 0.2 0.3 0.5 1.1 4.1 5 LaMarcus Aldridge C 36 BRK 47 12 22.3 5.4 9.7 0.55 0.3 1 0.304 5.1 8.8 0.578 0.566 1.9 2.2 0.873 1.6 3.9 5.5 0.9 0.3 1 0.9 1.7 12.9 6 Nickeil Alexander-Walker SG 23 TOT 65 21 22.6 3.9 10.5 0.372 1.6 5.2 0.311 2.3 5.3 0.433 0.449 1.2 1.7 0.743 0.6 2.3 2.9 2.4 0.7 0.4 1.4 1.6 10.6 6 Nickeil Alexander-Walker SG 23 NOP 50 19 26.3 4.7 12.6 0.375 1.9 6.1 0.311 2.8 6.5 0.434 0.45 1.4 1.9 0.722 0.7 2.6 3.3 2.8 0.8 0.4 1.7 1.8 12.8 6 Nickeil Alexander-Walker SG 23 UTA 15 2 9.9 1.1 3.2 0.333 0.7 2.2 0.303 0.4 1 0.4 0.438 0.7 0.8 0.917 0.1 1.5 1.5 1.1 0.3 0.3 0.5 1 3.5 7 Grayson Allen SG 26 MIL 66 61 27.3 3.9 8.6 0.448 2.4 5.9 0.409 1.5 2.7 0.533 0.588 1 1.1 0.865 0.5 2.9 3.4 1.5 0.7 0.3 0.7 1.5 11.1 8 Jarrett Allen C 23 CLE 56 56 32.3 6.6 9.7 0.677 0 0.2 0.1 6.6 9.6 0.688 0.678 2.9 4.2 0.708 3.4 7.3 10.8 1.6 0.8 1.3 1.7 1.7 16.1

In [0]:
# Importando os dados dos playoffs

# Localização e tipo do arquivo
new_file_location = "/FileStore/tables/2021_2022_NBA_Player_Stats___Playoffs.csv"
new_file_type = "csv"

# Parâmetros do arquivo .csv
infer_schema = "false"
first_row_is_header = "true"
delimiter = ","

# Criando dataframe dos playoffs
new_df = spark.read.format(new_file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", ";") \
  .load(new_file_location)

display(new_df.head(10))

Rk Player Pos Age Tm G GS MP FG FGA FG% 3P 3PA 3P% 2P 2PA 2P% eFG% FT FTA FT% ORB DRB TRB AST STL BLK TOV PF PTS 1 Precious Achiuwa C 22 TOR 6 1 27.8 4.2 8.7 0.481 0.8 2.7 0.313 3.3 6 0.556 0.529 1 1.7 0.6 1.3 3.5 4.8 1 0.2 0.8 1.5 2.3 10.2 2 Steven Adams C 28 MEM 7 5 16.3 1.3 3 0.429 0 0 0 1.3 3 0.429 0.429 0.9 1.6 0.545 2.1 4.3 6.4 2.1 0.1 0.1 0.6 1.7 3.4 3 Bam Adebayo C 24 MIA 18 18 34.1 5.8 9.7 0.594 0 0.1 0 5.8 9.7 0.598 0.594 3.2 4.2 0.763 2.1 5.9 8 2.7 1 0.7 2.1 3.1 14.8 4 Nickeil Alexander-Walker SG 23 UTA 1 0 5 2 2 1 0 0 0 2 2 1 1 1 1 1 0 1 1 1 1 0 0 0 5 5 Grayson Allen SG 26 MIL 12 5 25.4 3.1 6.8 0.451 1.6 4 0.396 1.5 2.8 0.529 0.567 0.6 0.9 0.636 0.4 2.5 2.9 1.3 0.7 0.3 0.8 1.8 8.3 6 Jose Alvarado PG 23 NOP 6 0 19.5 2.7 5.5 0.485 1 2.7 0.375 1.7 2.8 0.588 0.576 1.7 2.2 0.769 0.3 1 1.3 1.5 1.2 0.2 1.2 3 8 7 Kyle Anderson PF 28 MEM 12 1 18.4 2.4 4.3 0.569 0.3 1 0.25 2.2 3.3 0.667 0.598 0.9 1.5 0.611 0.8 3.5 4.3 1.8 0.9 0.6 0.8 1.8 6 8 Giannis Antetokounmpo PF 27 MIL 12 12 37.3 11.7 23.8 0.491 0.8 3.4 0.22 10.9 20.3 0.537 0.507 7.6 11.2 0.679 2.2 12 14.2 6.8 0.7 1.3 4.5 3.6 31.7 9 Thanasis Antetokounmpo SF 29 MIL 8 0 2.5 0.3 0.4 0.667 0 0 0 0.3 0.4 0.667 0.667 0.1 0.4 0.333 0.1 0.4 0.5 0.1 0.1 0 0 0.5 0.6 10 OG Anunoby SF 24 TOR 6 6 36.2 6.5 13.7 0.476 2.3 6.8 0.341 4.2 6.8 0.61 0.561 2 2.7 0.75 1.2 2.8 4 2.5 1 0.2 1.7 3 17.3

In [0]:
# Salvando os dados temporada regular na nuvem

# Salvar o DataFrame como uma tabela na plataforma Databricks
permanent_table_name = "2021_2022_NBA_Player_Stats___Regular_csv"

dbutils.fs.rm('dbfs:/user/hive/warehouse/2021_2022_nba_player_stats___regular_csv', recurse=True) # Limpando o diretório

df.write.format("delta").saveAsTable(permanent_table_name)

# Consultar a tabela criada no modelo flat
df_check = spark.sql(f"SELECT * FROM {permanent_table_name}")
display(df_check.limit(10))

Rk Player Pos Age Tm G GS MP FG FGA FG% 3P 3PA 3P% 2P 2PA 2P% eFG% FT FTA FT% ORB DRB TRB AST STL BLK TOV PF PTS 1 Precious Achiuwa C 22 TOR 73 28 23.6 3.6 8.3 0.439 0.8 2.1 0.359 2.9 6.1 0.468 0.486 1.1 1.8 0.595 2 4.5 6.5 1.1 0.5 0.6 1.2 2.1 9.1 2 Steven Adams C 28 MEM 76 75 26.3 2.8 5.1 0.547 0 0 0 2.8 5 0.548 0.547 1.4 2.6 0.543 4.6 5.4 10 3.4 0.9 0.8 1.5 2 6.9 3 Bam Adebayo C 24 MIA 56 56 32.6 7.3 13 0.557 0 0.1 0 7.3 12.9 0.562 0.557 4.6 6.1 0.753 2.4 7.6 10.1 3.4 1.4 0.8 2.6 3.1 19.1 4 Santi Aldama PF 21 MEM 32 0 11.3 1.7 4.1 0.402 0.2 1.5 0.125 1.5 2.6 0.56 0.424 0.6 1 0.625 1 1.7 2.7 0.7 0.2 0.3 0.5 1.1 4.1 5 LaMarcus Aldridge C 36 BRK 47 12 22.3 5.4 9.7 0.55 0.3 1 0.304 5.1 8.8 0.578 0.566 1.9 2.2 0.873 1.6 3.9 5.5 0.9 0.3 1 0.9 1.7 12.9 6 Nickeil Alexander-Walker SG 23 TOT 65 21 22.6 3.9 10.5 0.372 1.6 5.2 0.311 2.3 5.3 0.433 0.449 1.2 1.7 0.743 0.6 2.3 2.9 2.4 0.7 0.4 1.4 1.6 10.6 6 Nickeil Alexander-Walker SG 23 NOP 50 19 26.3 4.7 12.6 0.375 1.9 6.1 0.311 2.8 6.5 0.434 0.45 1.4 1.9 0.722 0.7 2.6 3.3 2.8 0.8 0.4 1.7 1.8 12.8 6 Nickeil Alexander-Walker SG 23 UTA 15 2 9.9 1.1 3.2 0.333 0.7 2.2 0.303 0.4 1 0.4 0.438 0.7 0.8 0.917 0.1 1.5 1.5 1.1 0.3 0.3 0.5 1 3.5 7 Grayson Allen SG 26 MIL 66 61 27.3 3.9 8.6 0.448 2.4 5.9 0.409 1.5 2.7 0.533 0.588 1 1.1 0.865 0.5 2.9 3.4 1.5 0.7 0.3 0.7 1.5 11.1 8 Jarrett Allen C 23 CLE 56 56 32.3 6.6 9.7 0.677 0 0.2 0.1 6.6 9.6 0.688 0.678 2.9 4.2 0.708 3.4 7.3 10.8 1.6 0.8 1.3 1.7 1.7 16.1

In [0]:
# Salvando os dados temporada de playoffs na nuvem

# Nome da nova tabela para dados dos playoffs
permanent_table_name_playoffs = "2021_2022_NBA_Player_Stats_Playoffs_csv"

# Limpar o diretório existente, se necessário
dbutils.fs.rm(f'dbfs:/user/hive/warehouse/{permanent_table_name_playoffs}', recurse=True)

# Salvar os dados dos playoffs como uma nova tabela Delta
new_df.write.format("delta").mode("overwrite").saveAsTable(permanent_table_name_playoffs)

# Verificar a nova tabela criada (também no modelo flat)
df_check_playoffs = spark.sql(f"SELECT * FROM {permanent_table_name_playoffs}")
display(df_check_playoffs.limit(10))


Rk Player Pos Age Tm G GS MP FG FGA FG% 3P 3PA 3P% 2P 2PA 2P% eFG% FT FTA FT% ORB DRB TRB AST STL BLK TOV PF PTS 1 Precious Achiuwa C 22 TOR 6 1 27.8 4.2 8.7 0.481 0.8 2.7 0.313 3.3 6 0.556 0.529 1 1.7 0.6 1.3 3.5 4.8 1 0.2 0.8 1.5 2.3 10.2 2 Steven Adams C 28 MEM 7 5 16.3 1.3 3 0.429 0 0 0 1.3 3 0.429 0.429 0.9 1.6 0.545 2.1 4.3 6.4 2.1 0.1 0.1 0.6 1.7 3.4 3 Bam Adebayo C 24 MIA 18 18 34.1 5.8 9.7 0.594 0 0.1 0 5.8 9.7 0.598 0.594 3.2 4.2 0.763 2.1 5.9 8 2.7 1 0.7 2.1 3.1 14.8 4 Nickeil Alexander-Walker SG 23 UTA 1 0 5 2 2 1 0 0 0 2 2 1 1 1 1 1 0 1 1 1 1 0 0 0 5 5 Grayson Allen SG 26 MIL 12 5 25.4 3.1 6.8 0.451 1.6 4 0.396 1.5 2.8 0.529 0.567 0.6 0.9 0.636 0.4 2.5 2.9 1.3 0.7 0.3 0.8 1.8 8.3 6 Jose Alvarado PG 23 NOP 6 0 19.5 2.7 5.5 0.485 1 2.7 0.375 1.7 2.8 0.588 0.576 1.7 2.2 0.769 0.3 1 1.3 1.5 1.2 0.2 1.2 3 8 7 Kyle Anderson PF 28 MEM 12 1 18.4 2.4 4.3 0.569 0.3 1 0.25 2.2 3.3 0.667 0.598 0.9 1.5 0.611 0.8 3.5 4.3 1.8 0.9 0.6 0.8 1.8 6 8 Giannis Antetokounmpo PF 27 MIL 12 12 37.3 11.7 23.8 0.491 0.8 3.4 0.22 10.9 20.3 0.537 0.507 7.6 11.2 0.679 2.2 12 14.2 6.8 0.7 1.3 4.5 3.6 31.7 9 Thanasis Antetokounmpo SF 29 MIL 8 0 2.5 0.3 0.4 0.667 0 0 0 0.3 0.4 0.667 0.667 0.1 0.4 0.333 0.1 0.4 0.5 0.1 0.1 0 0 0.5 0.6 10 OG Anunoby SF 24 TOR 6 6 36.2 6.5 13.7 0.476 2.3 6.8 0.341 4.2 6.8 0.61 0.561 2 2.7 0.75 1.2 2.8 4 2.5 1 0.2 1.7 3 17.3


Catálogo de Dados


Catálogo de Dados: NBA Player Stats 2021-2022

Descrição Geral dos Datasets
Nome dos Datasets:
  - Temporada Regular: '2021_2022_NBA_Player_Stats_Regular'
  - Playoffs: '2021_2022_NBA_Player_Stats_Playoffs'
- Fonte dos Dados: Dados coletados em https://www.kaggle.com/datasets/vivovinco/nba-player-stats/data e originalmente disponibilizados por Basketball reference (https://www.kaggle.com/datasets/vivovinco/nba-player-stats/data).
- Descrição dos Dados: Estatísticas detalhadas de jogadores da NBA durante a temporada regular e playoffs de 2021-2022.
- Período dos Dados: Temporada 2021-2022.

Detalhes das Colunas (aplicado a ambos os datasets)
- **Rk (Rank)**: Numérico. Representa o ranking do jogador. Valores de 1 a 500.
- **Player (Player's name)**: Texto. Nome completo do jogador.
- **Pos (Position)**: Categórico. Posição do jogador em campo (G - Guard, F - Forward, C - Center).
- **Age (Player's age)**: Numérico. Idade do jogador. Valores de 18 a 40.
- **Tm (Team)**: Categórico. Time de basquete ao qual o jogador pertence. Inclui todos os times da NBA.
- **G (Games played)**: Numérico. Número de jogos disputados na temporada. De 0 a 82.
- **GS (Games started)**: Numérico. Número de jogos que o jogador começou como titular. De 0 a 82.
- **MP (Minutes played per game)**: Numérico. Média de minutos jogados por partida. De 0 a 48.
- **FG (Field goals per game)**: Numérico. Média de arremessos de campo feitos por jogo. De 0 a 20.
- **FGA (Field goal attempts per game)**: Numérico. Média de tentativas de arremessos de campo por jogo. De 0 a 40.
- **FG% (Field goal percentage)**: Numérico. Percentual de acertos de arremessos de campo. De 0% a 100%.
- **3P (3-point field goals per game)**: Numérico. Média de arremessos de três pontos feitos por jogo. De 0 a 15.
- **3PA (3-point field goal attempts per game)**: Numérico. Média de tentativas de arremessos de três pontos por jogo. De 0 a 25.
- **3P% (3-point field goal percentage)**: Numérico. Percentual de acertos de três pontos. De 0% a 100%.
- **2P (2-point field goals per game)**: Numérico. Média de arremessos de dois pontos feitos por jogo. De 0 a 25.
- **2PA (2-point field goal attempts per game)**: Numérico. Média de tentativas de dois pontos por jogo. De 0 a 30.
- **2P% (2-point field goal percentage)**: Numérico. Percentual de acertos de dois pontos. De 0% a 100%.
- **eFG% (Effective field goal percentage)**: Numérico. Percentual de eficácia de arremesso ajustado para o valor de arremessos de três pontos. De 0% a 100%.
- **FT (Free throws per game)**: Numérico. Média de lances livres convertidos por jogo. De 0 a 20.
- **FTA (Free throw attempts per game)**: Numérico. Média de tentativas de lances livres por jogo. De 0 a 25.
- **FT% (Free throw percentage)**: Numérico. Percentual de acertos de lances livres. De 0% a 100%.
- **ORB (Offensive rebounds per game)**: Numérico. Média de rebotes ofensivos por jogo. De 0 a 15.
- **DRB (Defensive rebounds per game)**: Numérico. Média de rebotes defensivos por jogo. De 0 a 20.
- **TRB (Total rebounds per game)**: Numérico. Média total de rebotes por jogo. De 0 a 30.
- **AST (Assists per game)**: Numérico. Média de assistências por jogo. De 0 a 20.
- **STL (Steals per game)**: Numérico. Média de roubos de bola por jogo. De 0 a 10.
- **BLK (Blocks per game)**: Numérico. Média de bloqueios por jogo. De 0 a 10.
- **TOV (Turnovers per game)**: Numérico. Média de perdas de bola por jogo. De 0 a 10.
- **PF (Personal fouls per game)**: Numérico. Média de faltas pessoais por jogo. De 0 a 10.
- **PTS (Points per game)**: Numérico. Média de pontos por jogo. De 0 a 60.


Consultas SQL para responder as perguntas propostas


1) Correlação entre estatísticas ofensivas e defensivas

In [0]:
%sql

WITH 2021_2022_NBA_Player_Stats___Regular_csv AS (
  SELECT 
    Player,
    AVG(PTS) AS avg_points,
    AVG(AST) AS avg_assists,
    AVG(FGA) AS avg_field_goal_attempts,
    AVG(STL) AS avg_steals,
    AVG(BLK) AS avg_blocks
  FROM 
    2021_2022_NBA_Player_Stats___Regular_csv
  GROUP BY 
    Player
)

SELECT 
  CORR(avg_points, avg_steals) AS corr_points_steals,
  CORR(avg_points, avg_blocks) AS corr_points_blocks,
  CORR(avg_assists, avg_steals) AS corr_assists_steals,
  CORR(avg_assists, avg_blocks) AS corr_assists_blocks,
  CORR(avg_field_goal_attempts, avg_steals) AS corr_fga_steals,
  CORR(avg_field_goal_attempts, avg_blocks) AS corr_fga_blocks
FROM 
  2021_2022_NBA_Player_Stats___Regular_csv


corr_points_steals corr_points_blocks corr_assists_steals corr_assists_blocks corr_fga_steals corr_fga_blocks 0.6204161156836069 0.39773268505465803 0.6857376479520872 0.16967959053589055 0.639182126412356 0.33120803010025945


corr_points_steals (0.6204161156836069):

Existe uma correlação positiva moderadamente forte entre os pontos marcados e os roubos de bola. Isso sugere que jogadores que pontuam mais também tendem a realizar mais roubos de bola. Isso pode indicar alta atividade em quadra, tanto ofensiva quanto defensivamente.

corr_points_blocks (0.39773268505465803):

Há uma correlação positiva moderada entre os pontos marcados e os bloqueios. Isso indica que jogadores que pontuam mais também podem contribuir com bloqueios, sugerindo que alguns atletas efetivos no ataque também participam ativamente na defesa, especialmente aqueles que jogam perto do garrafão.

corr_assists_steals (0.6857376479520872):

Esta é a correlação mais forte entre as medidas analisadas, mostrando uma forte relação positiva entre assistências e roubos de bola. Isso reforça a ideia de que jogadores com boa visão de jogo e habilidades de passe (que geralmente resultam em assistências) também são ativos e eficazes na defesa, demonstrando reflexos rápidos e boa leitura das jogadas adversárias.

corr_assists_blocks (0.16967959053589055):

A correlação entre assistências e bloqueios é relativamente baixa, sugerindo que não há uma relação forte entre ser um bom passador e realizar bloqueios. Isso pode refletir o fato de que essas habilidades são geralmente mais específicas para diferentes tipos de jogadores e posições em quadra.

corr_fga_steals (0.639182126412356):

Existe uma correlação positiva moderadamente forte entre o número de arremessos tentados e roubos de bola. Isso pode indicar que jogadores que são mais ofensivamente agressivos (tentam mais arremessos) também são ativos defensivamente, envolvendo-se em jogadas que podem resultar em roubos de bola.

corr_fga_blocks (0.33120803010025945):

Há uma correlação moderada entre arremessos tentados e bloqueios. Isso sugere que, embora menos forte do que outras correlações, jogadores que tentam muitos arremessos podem também contribuir com bloqueios, indicando um papel ativo em ambas as extremidades da quadra.

In [0]:
%sql

WITH 2021_2022_NBA_Player_Stats_Playoffs_csv AS (
  SELECT 
    Player,
    AVG(PTS) AS avg_points,
    AVG(AST) AS avg_assists,
    AVG(FGA) AS avg_field_goal_attempts,
    AVG(STL) AS avg_steals,
    AVG(BLK) AS avg_blocks
  FROM 
    2021_2022_NBA_Player_Stats_Playoffs_csv
  GROUP BY 
    Player
)

SELECT 
  CORR(avg_points, avg_steals) AS corr_points_steals,
  CORR(avg_points, avg_blocks) AS corr_points_blocks,
  CORR(avg_assists, avg_steals) AS corr_assists_steals,
  CORR(avg_assists, avg_blocks) AS corr_assists_blocks,
  CORR(avg_field_goal_attempts, avg_steals) AS corr_fga_steals,
  CORR(avg_field_goal_attempts, avg_blocks) AS corr_fga_blocks
FROM 
  2021_2022_NBA_Player_Stats_Playoffs_csv


corr_points_steals corr_points_blocks corr_assists_steals corr_assists_blocks corr_fga_steals corr_fga_blocks 0.6235966764128418 0.461033950853624 0.6776032890324655 0.2690861572064153 0.6411462072970808 0.4159373039062268


Pontos e Roubos (Steals)

Temporada Regular: A correlação era significativa, sugerindo que jogadores que pontuam mais também participam ativamente na defesa.
Playoffs: A correlação permaneceu forte (0.6235966764128418), indicando que essa tendência se intensifica nos playoffs.

Pontos e Bloqueios (Blocks)

Temporada Regular: Existia uma correlação moderada, indicando que jogadores que pontuam também contribuem com bloqueios.
Playoffs: A correlação ainda é moderada (0.461033950853624), mas ligeiramente mais fraca comparada à temporada regular.

Assistências e Roubos

Temporada Regular: A correlação era forte, refletindo que jogadores com boa visão de jogo e habilidades de passe também são bons na defesa.
Playoffs: A correlação ficou ainda mais forte (0.6776032890324655), reforçando que a capacidade de antecipação e atenção à defesa são críticas nos playoffs.

Assistências e Bloqueios

Temporada Regular: A correlação era relativamente fraca.
Playoffs: A correlação aumentou um pouco (0.2690861572064153), mas ainda é considerada fraca, sugerindo que essa não é uma tendência generalizada.

Arremessos Tentados (FGA) e Roubos

Temporada Regular: Correlação moderadamente forte, mostrando que jogadores que tentam muitos arremessos também são envolvidos na defesa.
Playoffs: A correlação é semelhante (0.6411462072970808), mantendo a ideia de que atividade ofensiva elevada acompanha participação defensiva.

Arremessos Tentados e Bloqueios

Temporada Regular: A correlação era moderada.
Playoffs: Similarmente moderada (0.4159373039062268), indicando consistência nesse aspecto entre temporada regular e playoffs.

Conclusão Geral da Comparação
Os dados dos playoffs mostram uma tendência geral de manutenção ou mesmo intensificação das correlações observadas durante a temporada regular. As habilidades que combinam ação ofensiva com defesa ativa parecem ainda mais pronunciadas nos playoffs, talvez devido à maior intensidade e aos maiores desafios defensivos que os jogos eliminatórios apresentam. Jogadores que conseguem manter ou elevar seu desempenho em ambas as extremidades da quadra são especialmente valorizados durante os playoffs, onde cada posse de bola é crucial.


2) Jogadores que se destacam tanto ofensivamente quanto defensivamente


Para identificar quais jogadores se destacam tanto ofensivamente quanto defensivamente na NBA, deve-se considerar jogadores que estão acima de certos percentis em estatísticas ofensivas (como pontos e assistências) e defensivas (como roubos de bola e bloqueios). A consulta SQL a seguir calcula essas estatísticas e lista os jogadores que estão no top 5% (fase regular) e top 10% (playoffs) em pelo menos duas dessas categoria. Trata-se de um grupo bem seleto de jogadores de alto valor agregado para suas respectivas equipes.



Lista de destaques fase regular

In [0]:
%sql

WITH 2021_2022_NBA_Player_Stats___Regular_csv AS (
    SELECT 
        Player,
        AVG(PTS) AS avg_points,
        AVG(AST) AS avg_assists,
        AVG(STL) AS avg_steals,
        AVG(BLK) AS avg_blocks
    FROM 
        2021_2022_NBA_Player_Stats___Regular_csv
    GROUP BY 
        Player
),
Percentiles AS (
    SELECT 
        Player,
        avg_points,
        avg_assists,
        avg_steals,
        avg_blocks,
        PERCENT_RANK() OVER (ORDER BY avg_points DESC) AS pr_points,
        PERCENT_RANK() OVER (ORDER BY avg_assists DESC) AS pr_assists,
        PERCENT_RANK() OVER (ORDER BY avg_steals DESC) AS pr_steals,
        PERCENT_RANK() OVER (ORDER BY avg_blocks DESC) AS pr_blocks
    FROM 
        2021_2022_NBA_Player_Stats___Regular_csv
)
SELECT 
    Player,
    avg_points,
    avg_assists,
    avg_steals,
    avg_blocks
FROM 
    Percentiles
WHERE 
    (pr_points <= 0.05 AND pr_steals <= 0.05) OR 
    (pr_points <= 0.05 AND pr_blocks <= 0.05) OR 
    (pr_assists <= 0.05 AND pr_steals <= 0.05) OR 
    (pr_assists <= 0.05 AND pr_blocks <= 0.05)


Player avg_points avg_assists avg_steals avg_blocks Joel Embiid 30.6 4.2 1.1 1.5 LeBron James 30.3 6.2 1.3 1.1 Giannis Antetokounmpo 29.9 5.8 1.1 1.4 Kyrie Irving 27.4 5.8 1.4 0.6 Nikola Joki? 27.1 7.9 1.5 0.9 Donovan Mitchell 25.9 5.3 1.5 0.2 Stephen Curry 25.5 6.3 1.3 0.4 Karl-Anthony Towns 24.6 3.6 1.0 1.1 Shai Gilgeous-Alexander 24.5 5.9 1.3 0.8 Paul George 24.3 5.7 2.2 0.4 Anthony Davis 23.2 3.1 1.2 2.3 Pascal Siakam 22.8 5.3 1.3 0.6 Darius Garland 21.7 8.6 1.3 0.1 Jimmy Butler 21.4 5.5 1.6 0.5 Anthony Edwards 21.3 3.8 1.5 0.6 Dejounte Murray 21.1 9.2 2.0 0.3 Fred VanVleet 20.3 6.7 1.7 0.5 LaMelo Ball 20.1 7.6 1.6 0.4 Jrue Holiday 18.3 6.8 1.6 0.4 Tyrese Haliburton 15.700000000000001 8.4 1.7333333333333334 0.6333333333333333 Chris Paul 14.7 10.8 1.9 0.3 Ricky Rubio 13.1 6.6 1.4 0.2 Marcus Smart 12.1 5.9 1.7 0.3 Zavier Simpson 11.0 7.5 1.3 1.0 Draymond Green 7.5 7.0 1.3 1.1 Derrick Walton 6.3 7.0 2.3 1.3


Lista de destaques fase playoffs

In [0]:
%sql

WITH 2021_2022_NBA_Player_Stats_Playoffs_csv AS (
    SELECT 
        Player,
        AVG(PTS) AS avg_points,
        AVG(AST) AS avg_assists,
        AVG(STL) AS avg_steals,
        AVG(BLK) AS avg_blocks
    FROM 
        2021_2022_NBA_Player_Stats_Playoffs_csv
    GROUP BY 
        Player
),
Percentiles AS (
    SELECT 
        Player,
        avg_points,
        avg_assists,
        avg_steals,
        avg_blocks,
        PERCENT_RANK() OVER (ORDER BY avg_points DESC) AS pr_points,
        PERCENT_RANK() OVER (ORDER BY avg_assists DESC) AS pr_assists,
        PERCENT_RANK() OVER (ORDER BY avg_steals DESC) AS pr_steals,
        PERCENT_RANK() OVER (ORDER BY avg_blocks DESC) AS pr_blocks
    FROM 
        2021_2022_NBA_Player_Stats_Playoffs_csv
)
SELECT 
    Player,
    avg_points,
    avg_assists,
    avg_steals,
    avg_blocks
FROM 
    Percentiles
WHERE 
    (pr_points <= 0.1 AND pr_steals <= 0.1) OR 
    (pr_points <= 0.1 AND pr_blocks <= 0.1) OR 
    (pr_assists <= 0.1 AND pr_steals <= 0.1) OR 
    (pr_assists <= 0.1 AND pr_blocks <= 0.1)


Player avg_points avg_assists avg_steals avg_blocks Giannis Antetokounmpo 31.7 6.8 0.7 1.3 Luka Don?i? 31.7 6.4 1.8 0.6 Nikola Joki? 31.0 5.8 1.6 1.0 Jimmy Butler 27.4 4.6 2.1 0.6 Stephen Curry 27.4 5.9 1.3 0.4 Ja Morant 27.1 9.8 2.0 0.4 Jayson Tatum 25.6 6.2 1.2 0.9 Anthony Edwards 25.2 3.0 1.2 1.2 Pascal Siakam 22.8 5.8 1.2 1.0 Karl-Anthony Towns 21.8 2.2 0.7 2.0 Kyrie Irving 21.3 5.3 1.8 1.3 DeMar DeRozan 20.8 4.8 1.8 0.4 Jrue Holiday 19.1 6.5 1.8 0.6 Chris Paul 17.5 8.3 1.5 0.2 Marcus Smart 15.4 5.9 1.2 0.2 Khris Middleton 14.5 7.0 1.5 0.0 Monte Morris 14.0 5.4 1.2 0.0 Fred VanVleet 13.8 6.3 1.8 1.0 D'Angelo Russell 12.0 6.7 1.5 0.0 Draymond Green 8.0 6.3 1.1 1.0


3) Relação entre idade e desempenho na NBA 2021-2022



Fase regular

In [0]:
%sql

WITH AgeStats AS (
    SELECT 
        Age,
        AVG(PTS) AS avg_points,
        AVG(AST) AS avg_assists,
        AVG(TRB) AS avg_rebounds,
        AVG(STL) AS avg_steals,
        AVG(BLK) AS avg_blocks
    FROM 
        2021_2022_NBA_Player_Stats___Regular_csv
    GROUP BY 
        Age
)
SELECT 
    CORR(Age, avg_points) AS correlation_points_age,
    CORR(Age, avg_assists) AS correlation_assists_age,
    CORR(Age, avg_rebounds) AS correlation_rebounds_age,
    CORR(Age, avg_steals) AS correlation_steals_age,
    CORR(Age, avg_blocks) AS correlation_blocks_age
FROM 
    AgeStats


correlation_points_age correlation_assists_age correlation_rebounds_age correlation_steals_age correlation_blocks_age -0.05542464420349225 0.13178364167040066 -0.09245079126384152 -0.12071302383358612 0.025081532513309388


Fase playoffs

In [0]:
%sql

WITH AgeStats AS (
    SELECT 
        Age,
        AVG(PTS) AS avg_points,
        AVG(AST) AS avg_assists,
        AVG(TRB) AS avg_rebounds,
        AVG(STL) AS avg_steals,
        AVG(BLK) AS avg_blocks
    FROM 
        2021_2022_NBA_Player_Stats_Playoffs_csv
    GROUP BY 
        Age
)
SELECT 
    CORR(Age, avg_points) AS correlation_points_age,
    CORR(Age, avg_assists) AS correlation_assists_age,
    CORR(Age, avg_rebounds) AS correlation_rebounds_age,
    CORR(Age, avg_steals) AS correlation_steals_age,
    CORR(Age, avg_blocks) AS correlation_blocks_age
FROM 
    AgeStats


correlation_points_age correlation_assists_age correlation_rebounds_age correlation_steals_age correlation_blocks_age 5.698857266094467E-4 0.5812869445134746 0.246179021060128 0.11972125153849586 0.1470290561824736


Os resultados das correlações entre as estatísticas dos jogadores e suas idades durante os playoffs mostram algumas diferenças interessantes em comparação com os resultados da temporada regular. Vamos interpretar cada correlação dos dados dos playoffs e comparar com os resultados da fase regular:

Pontos e Idade (0.0005698857266094467):

Playoffs: A correlação entre pontos e idade é praticamente zero, indicando que não há relação discernível entre a idade dos jogadores e a quantidade de pontos que marcam durante os playoffs.
Temporada Regular: A correlação também era muito fraca e negativa, sugerindo uma tendência quase inexistente.
Comparação: Em ambos os casos, a idade não parece ser um fator determinante para a pontuação dos jogadores.

Assistências e Idade (0.5812869445134746):

Playoffs: Existe uma correlação positiva moderada a forte, indicando que jogadores mais velhos tendem a ter mais assistências nos playoffs. Isso pode refletir maior experiência ou um papel tático mais centrado em distribuição do jogo.
Temporada Regular: A correlação era positiva, mas mais fraca.
Comparação: A influência da idade sobre as assistências é mais pronunciada nos playoffs do que na temporada regular, possivelmente devido à natureza de alta pressão dos jogos que valoriza a experiência.

Rebotes e Idade (0.246179021060128):

Playoffs: A correlação positiva fraca a moderada sugere que jogadores mais velhos podem ter um desempenho ligeiramente melhor nos rebotes durante os playoffs.
Temporada Regular: A correlação era negativa e fraca.
Comparação: Diferente da temporada regular, onde a idade parecia levar a um declínio nos rebotes, nos playoffs, a experiência pode estar ajudando os jogadores a serem mais efetivos nesse aspecto.

Roubos de Bola e Idade (0.11972125153849586):

Playoffs: A correlação é positiva, mas fraca, indicando uma leve tendência de jogadores mais velhos conseguirem mais roubos de bola.
Temporada Regular: A correlação era negativa e fraca.
Comparação: Similar aos rebotes, jogadores mais velhos parecem se sair ligeiramente melhor em roubos de bola durante os playoffs comparados à temporada regular.

Bloqueios e Idade (0.1470290561824736):

Playoffs: Correlação positiva fraca, sugerindo que jogadores mais velhos podem bloquear ligeiramente mais chutes nos playoffs.
Temporada Regular: A correlação era praticamente inexistente.
Comparação: A tendência para bloqueios parece aumentar levemente com a idade nos playoffs, diferentemente do observado na temporada regular.

Conclusão Geral
Os resultados indicam que, durante os playoffs, a experiência (idade) dos jogadores pode ter uma influência mais positiva em assistências, rebotes, roubos de bola e bloqueios, em comparação com a temporada regular. Isso sugere que as habilidades e o conhecimento adquiridos com a idade podem ser mais valiosos sob a pressão dos playoffs, onde cada jogada é crucial.


Carga no Delta Lake


Neste projeto, optou-se por uma carga incremental. Ambos os datasets são extraídos do banco de dados e a transformação executada é simplesmente uma combinação (merge), de forma que possam ser feitas futuras análises de dados em cima de toda a temporada 2021 - 2022 (fase regular + playoffs)

In [0]:
# Extração

# Carregando as tabelas delta em dataframes novos

# Carregar os dados da temporada regular
df_regular_season = spark.read.format("delta").table("2021_2022_NBA_Player_Stats___Regular_csv")

# Carregar os dados dos playoffs
df_playoffs = spark.read.format("delta").table("2021_2022_NBA_Player_Stats_Playoffs_csv")



In [0]:
# Transformação e carga
from delta.tables import DeltaTable

# Obter o objeto DeltaTable para a tabela da temporada regular
delta_table_regular = DeltaTable.forPath(spark, "dbfs:/user/hive/warehouse/2021_2022_nba_player_stats___regular_csv")

# Comando de merge para combinar dados dos playoffs na tabela da temporada regular usando apenas 'Player' como chave
(delta_table_regular.alias("regular")
 .merge(
     df_playoffs.alias("playoffs"),
     "regular.Player = playoffs.Player")  # Usando apenas Player como chave
 .whenMatchedUpdateAll()  # Atualiza os dados existentes se houver correspondência
 .whenNotMatchedInsertAll()  # Insere novos dados dos playoffs se não houver correspondência na tabela regular
 .execute()
)


In [0]:
# Verificação

# Carregar e mostrar os dados atualizados da tabela da temporada regular incluindo playoffs
updated_data = spark.read.format("delta").table("2021_2022_NBA_Player_Stats___Regular_csv")
updated_data.show()


+---+-----------------+---+---+---+---+---+----+---+----+-----+---+---+-----+---+---+-----+-----+---+---+-----+---+---+----+---+---+---+---+---+----+
| Rk|           Player|Pos|Age| Tm|  G| GS|  MP| FG| FGA|  FG%| 3P|3PA|  3P%| 2P|2PA|  2P%| eFG%| FT|FTA|  FT%|ORB|DRB| TRB|AST|STL|BLK|TOV| PF| PTS|
+---+-----------------+---+---+---+---+---+----+---+----+-----+---+---+-----+---+---+-----+-----+---+---+-----+---+---+----+---+---+---+---+---+----+
|  4|     Santi Aldama| PF| 21|MEM| 32|  0|11.3|1.7| 4.1|0.402|0.2|1.5|0.125|1.5|2.6| 0.56|0.424|0.6|  1|0.625|  1|1.7| 2.7|0.7|0.2|0.3|0.5|1.1| 4.1|
|  5|LaMarcus Aldridge|  C| 36|BRK| 47| 12|22.3|5.4| 9.7| 0.55|0.3|  1|0.304|5.1|8.8|0.578|0.566|1.9|2.2|0.873|1.6|3.9| 5.5|0.9|0.3|  1|0.9|1.7|12.9|
|  8|    Jarrett Allen|  C| 23|CLE| 56| 56|32.3|6.6| 9.7|0.677|  0|0.2|  0.1|6.6|9.6|0.688|0.678|2.9|4.2|0.708|3.4|7.3|10.8|1.6|0.8|1.3|1.7|1.7|16.1|
| 10|  Justin Anderson| SF| 28|TOT| 16|  6|19.8|2.3| 5.9|0.379|0.9|3.7|0.254|1.3|2.3|0.583|0.458|0.9


4) Os três melhores jogadores da liga na temporada completa 


Avaliando os três melhores jogadores do campeonato com os dados atualizados da temporada completa (regular + playoffs)

In [0]:
%sql

SELECT 
    Player,
    Tm as Team,
    AVG(PTS) as Avg_Points,
    AVG(TRB) as Avg_Rebounds,
    AVG(AST) as Avg_Assists,
    (AVG(PTS) + AVG(TRB) + AVG(AST)) as Efficiency_Score
FROM 
    2021_2022_NBA_Player_Stats___Regular_csv
GROUP BY 
    Player, Tm
ORDER BY 
    Efficiency_Score DESC
LIMIT 3;


Player Team Avg_Points Avg_Rebounds Avg_Assists Efficiency_Score Giannis Antetokounmpo MIL 31.7 14.2 6.8 52.699999999999996 Nikola Joki? DEN 31.0 13.2 5.8 50.0 Luka Don?i? DAL 31.7 9.8 6.4 47.9


Os jogadores mais decisivos da temporada completa (somando pontos, rebotes e assistências por jogo) são em ordem Giannis Antetokounpo, Nikola Jokic e Luka Doncic. 

Na temporada NBA 2021-2022, Nikola Jokic do Denver Nuggets foi eleito o MVP da liga. Jokic liderou os Nuggets a um recorde de 48-34, apesar de ausências importantes na equipe. O Milwaukee Bucks, com Giannis Antetokounmpo, e o Philadelphia 76ers, com Joel Embiid, ambos terminaram a temporada regular com registros de 51-31. Ambos os jogadores também tiveram performances notáveis e foram finalistas para o prêmio MVP junto com Jokic. Luka Dončić teve uma temporada notável em 2021-2022, levando o Dallas Mavericks a um bom desempenho na liga. Ele continuou sendo um dos principais jogadores da NBA, demonstrando excelente habilidade em pontuação, assistências e rebotes. Sua performance geral o colocou entre os candidatos ao prêmio de MVP, embora não tenha ganho, ele ainda recebeu votos significativos na votação, refletindo seu impacto e valor para sua equipe (Sporting News)​.

Esses resultados indicam um alto nível de desempenho e contribuição dos jogadores mencionados para seus respectivos times, refletindo diretamente na avaliação de suas eficácias individuais em relação às estatísticas de pontos, rebotes e assistências discutidas anteriormente.

Curiosamente, nenhum dos jogadores citados foi campeão, pois o vencedor da NBA 2021-2022 foi o Golden State Warriors.



Análise e autoavaliação


Qualidade dos dados 

Os datasets originais estão completamente curados e já bem tratados quando disponibilizados. Uma limitação que não foi abordada neste projeto é que algum dos jogadores são listados mais uma de vez no arquivo .csv da fase regular. Isso acontece pois jogaram a liga por mais de uma equipe devido a janela de transferências. A abordagem correta a se fazer era estabelecer um pré-processamento de dados capaz de combinar as entradas do mesmo jogador em uma só. Como o tratamento de dados foi o foco do módulo passado, essa etapa foi ignorada. Outro pré-processamento menos relevante era corrigir os caracteres especiais dos nomes dos jogadores estrangeiros como por exemplo Luka Dončić ao invés de Luka Don?i?. De qualquer forma, a maior parte dos extensos dataset cumpriu o seu papel no objetivo proposto por esse projeto.

Solução do problema e autoavaliação

Apesar da limitação mencionada acima, todas as perguntas elaboradas na seção de objetivos foram respondidas com sucesso. Em relação aos dados, foi realizado todo o procedimento de coleta, importação, armazenamento na nuvem. Utilizou-se o modelo flat e construiu-se um catálogo de dados detalhado. Por fim, uma carga foi imposta de maneira a combinar os resultados de toda a liga em um só data lake. Desta forma, a autoavaliação é bem positiva no sentido que os objetivos foram alcançados com sucesso. Trabalhos futuros envolvem a correção de jogadores duplicados no banco de dados e trazer novos tipos de análises estatíticas em torno dos dados disponíveis. Além disso, pode-se sempre ampliar o banco de dados com os dados das temporadas seguintes. 